### PR TITLE
Do not require a password on the truststore JKS

### PR DESCRIPTION
**Pull request by joshelser**
A password on a truststore provides no security value, only validation
that the JKS is the JKS that the user expects.

Signed-off-by: Josh Elser <elserj@apache.org>

### CODE REVIEW

**Comment by ctubbsii** [comment]
tl;dr I think this is a bad idea (see my case below). However, I'm open to being convinced... Is there a some use case which cannot be satisfied with the current checks? What's driving this?

****

While it may not be necessary to provide confidentiality of the truststore, "validation that the JKS is the JKS that the user expects" is the very definition of integrity. Passphrases on java keystores (including truststores) provide both confidentiality and integrity. So the log message is incorrect: it *is* providing security value, by enforcing the use of integrity beyond file system permissions.

While we don't necessarily need to enforce the use of integrity measures by requiring a passphrase (we could leave that to the user), I think it's better that we do... it encourages good security practices, and ensures that they must take active measures to select less integrity (by choosing a weak and well-known password like "changeit"). If we do what you propose, the default posture for the user will be shifted from one of having integrity and needing to take steps to reduce integrity, to one of not having integrity and needing to take action to provide integrity. I'm inclined to prefer the former, which is what we have now.

**Comment by ctubbsii** [comment]
I actually think there might be other issues with this boolean, though... like the fact that we assume if not everything is set, then we didn't want HTTPS. However, I think we should check if *any* item is non-empty, and if if any is non-empty, then assume we do want HTTPS, and then check that we have everything else we need and fail if we don't rather than fall back to insecure HTTP.

If we change that logic a bit, then it might make sense to consider the passphrases optional (but I'd still recommend to set them).

Another problem I noticed... I'm not sure we're checking for CLIENT_AUTH... we should be able to set that to lock down the monitor to only trusted users. Maybe I just didn't see it... I thought it was there, but I could be wrong.

**Comment by joshelser** [comment]
> While it may not be necessary to provide confidentiality of the truststore, "validation that the JKS is the JKS that the user expects" is the very definition of integrity.

Yes, thanks for re-defining integrity for me ;). Thought it was obvious that's what I was describing. No security, just integrity of the JKS.

> So the log message is incorrect: it is providing security value, by enforcing the use of integrity beyond file system permissions.

Truststores, in this context, do not contain private material, thus there is no need to enforce confidentiality -- in fact, the typical deployment model for Java installations has a globally readable truststore.

**Comment by ctubbsii** [comment]
> Thought it was obvious that's what I was describing. No security, just integrity of the JKS.

It wasn't obvious because I was/am confused by your continued phrasing which implies integrity is not security.

> Truststores, in this context, do not contain private material, thus there is no need to enforce confidentiality

I agree and said as much myself. Truststores don't need confidentiality. They do need integrity. Passphrases provide both, but truststores only need the latter.

**Comment by joshelser** [comment]
> overlooks integrity as a type of security.

Integrity is a part of security, not "security" if you ask me. If I only give you integrity, it is not "secure". That is what I'm implying.

**Comment by ctubbsii** [comment]
> Integrity is a part of security, not "security" if you ask me. If I only give you integrity, it is not "secure". That is what I'm implying.

Okay. That's not how I would use those terms, but I now understand what you were saying. Thanks for clarifying. Hopefully you understood where I was coming from also (using integrity as a type of security, and not a component of security).

Semantics of "security" now hopefully resolved, I'm still concerned about the removal of an enforcement of an integrity "best-practice". What is the use case/driving force for removing this enforcement?

**Comment by joshelser** [comment]
> What is the use case/driving force for removing this enforcement?

I have a truststore which doesn't have a password and I cannot configure Accumulo to use SSL. Cut and dry.

The other things you suggested seem like fine other things that could be pursued.

**Comment by ctubbsii** [comment]
> I have a truststore which doesn't have a password

Can't a password be set on the truststore to add integrity to it? That seems better (and easier) than changing the code base to relax the integrity constraint.

**Comment by joshelser** [comment]
> That seems better (and easier) than changing the code base to relax the integrity constraint.

I disagree.

**Comment by keith-turner** [comment]
I was lurking around in the Accumulo monitor code and Jetty code to understand this.  One thing I noticed in the Jetty code is that if the truststore is not set, then it will use the keystore as the truststore.  So one possible option is we require both truststore path and password are set OR no truststore props are set.  When nothing is set for the truststore, then we don't set anything on Jetty's SslConnectionFactory for truststore and it falls back to the keystore.

I think an issue needs to be opened about the monitor falling back to http if https config is present but incorrect.  The monitor should probably fail to start.

**Comment by joshelser** [comment]
> So one possible option is we require both truststore path and password are set OR no truststore props are set. When nothing is set for the truststore, then we don't set anything on Jetty's SslConnectionFactory for truststore and it falls back to the keystore.

I'm not following how this helps my case, Keith. I have a truststore with no password. I don't want the keystore to be used.

**Comment by ctubbsii** [comment]
> I disagree.

Okay... so ease is subjective and we disagree on which is easier. Either way, I'm still concerned about potentially relaxing an integrity constraint without sufficient justification, even if it is slightly easier.

You've requested my opinion, and my opinion is that stores without this integrity check put users at greater risk and we shouldn't let users shoot themselves in the foot like that. I don't think this risk justifies whatever ease of use comes from avoiding setting a password to add an integrity check to the store (which can be done by running the simple command `keytool -storepasswd -keystore my.truststore`). Even if the risk is small, I think the ease of use benefit is smaller.

I'm not convinced that you've made a case that this ease does justify the risks, but I'm willing to hear you out if you wish to make such a case. Otherwise, feel free to proceed with my objection outstanding. I won't veto or block it... I just think it's a bad idea. Even if you do merge this change, I will continue to recommend people use passphrases on their truststores, for whatever extra integrity checking it enables.

**Comment by keith-turner** [comment]
> I'm not following how this helps my case, Keith. I have a truststore with no password. I don't want the keystore to be used.

Ok.  I suggested because I suspected the truststore was created because it was required and duplicated CA info already in the keystore.  But that is not your use case, so ignore that line of thought. 

I think there two distinct issues to consider with this change : password quality and was a password set.  It may make sense to require setting a password.  However, nothing should be baked into the code that attempts to measure password quality and reject bad passwords.  If we are going to disallow empty string, then why stop there?  Why not also disallow single character passwords?  

In the 2.0 branch I think I added something to AccumuloConfig to check if a prop was set or not, but I don't think this is in 1.9.  If there is an easy way to distinguish between set and not set in 1.9, then I think requiring the user to set any password (including empty string) would be good.  If this solution is not easy to do in 1.9, then I don't have strong opinion for or against making this change as is.

**Comment by joshelser** [comment]
> Ok. I suggested because I suspected the truststore was created because it was required and duplicated CA info already in the keystore. But that is not your use case, so ignore that line of thought.

Gotcha. No, that's not the case here.

**Comment by joshelser** [comment]
> I think there two distinct issues to consider with this change : password quality and was a password set.

This strikes me as a future improvement. Mine is specifically targeted to this one scenario which prevents me from enabling TLS. Your suggestion, as well as Christopher's, sound like good other improvements to consider.

**Comment by ctubbsii** [comment]
There's currently no way to distinguish between set empty and unset, as is. Disallowing an empty passphrase is consistent with the behavior of `keytool`, which requires a passphrase of at least 6 characters. If we were to allow it, we should at least spit out a warning that we cannot verify its integrity.

**Comment by joshelser** [comment]
> If we were to allow it, we should at least spit out a warning that we cannot verify its integrity.

How about 086268b @ctubbsii ?

**Comment by ctubbsii** [comment]
> How about [086268b](https://github.com/apache/accumulo/commit/086268b1915fcef9e2c6d1d984db6449d7e17b53) @ctubbsii ?

I still think it's a bad idea to encourage this, but I'm more comfortable with the warning being present.

**Comment by joshelser** [comment]
PS, I'm still fighting a deterministic unit test failure in TServerUtilsTest where the original port grabbed doesn't actually preclude the second from grabbing the same...

**Comment by joshelser** [comment]
> ServerSocket[addr=0.0.0.0/0.0.0.0,localport=1024]

Seems like the newer Thrift is to blame, or just TServerUtilsTest. We only pass in the port at TServerUtils.java:279 when constructing the `TServerSocket` which results in the test not binding to localhost as expected. I think this is why it doesn't fail.

How doesn't this fail for everyone though?

**Comment by joshelser** [comment]
Well, at this point, this is obviously not something related to this change. Will spin this out.

**Comment by ctubbsii** [comment]
I think this was cherry-picked from 1.9 to 2.0. Typically, we apply to the oldest branch and merge forward to newer branches, to ensure whatever we've patched in older releases aren't regressed in newer ones.

**Comment by joshelser** [comment]
Crap, I forgot about that. Will push a merge.

**Comment by joshelser** [comment]
Nevermind, you did this.

**Comment by ctubbsii** [comment]
Yeah, not a big deal to fix. I just wanted to mention it to serve a reminder (especially to any newcomer observers). :smiley_cat:
